fe/fe-core/src/main/java/org/apache/doris/common/MetaWriter.java

```diff
@@ -65,67 +67,70 @@
 public class MetaWriter {
     private static final Logger LOG = LogManager.getLogger(MetaWriter.class);
 
+    public static MetaWriter writer = new MetaWriter();
+
+    private interface Delegate {
+        long doWork(String name, WriteMethod method) throws IOException;
+    }
+
+    private interface WriteMethod {
+        long write() throws IOException;
+    }
+
+    private Delegate delegate;
+
+    public void setDelegate(CountingDataOutputStream dos, List<MetaIndex> indices) {
+        this.delegate = (name, method) -> {
+            indices.add(new MetaIndex(name, dos.getCount()));
+            return method.write();
+        };
+    }
+
+    public long doWork(String name, WriteMethod method) throws IOException {
+        if(delegate == null){
+            return method.write();
+        }
+        return delegate.doWork(name, method);
+    }
+
     public static void write(File imageFile, Catalog catalog) throws IOException {
         // save image does not need any lock. because only checkpoint thread will call this method.
         LOG.info("start save image to {}. is ckpt: {}", imageFile.getAbsolutePath(), Catalog.isCheckpointThread());
 
-        long checksum = 0;
+        final AtomicReference<Long> checksum = new AtomicReference<>(0L);
```

Comment:
why use AtomicReference<Long> instead of long here?

Reply:
Because the anonymous function in lambda expression only support final argument.
So when you write like this:

```
long checksum = 0;
checksum = MetaWriter.writer.doWork("header", () -> catalog.saveHeader(dos, 1000L, checksum));
```

It will report a compilation error.

Reply:
I think it is not related to `AtomicReference`, it is about `final`. In lamda express, it should be `final`

Reply:
Yes. It's not about concurrent problem but final. I used Reference instead of AtomicReference.